PyRoute/TradeCalculation.py

```diff
@@ -756,6 +783,25 @@ def route_weight(self, star, target):
         weight -= star.importance + target.importance
         return weight
 
+    def calculate_components(self):
```

Comment:
This is a great idea, but, I think, in the wrong place. 

You need to add the call to `calculate_components()` to the `generate_routes()` before the call to `generate_base_routes()`. Then in the function `base_route_filter()` add the check for star/neighbor in the same component. Rather than running through the complete set of BTNs another time. 

Because of the ordering of the calls the setting of `redzone` in this function isn't used.

Reply:
@tjoneslo , thanks for your feedback.

`calculate_components()` needs to run **after** the edges in galaxy.stars are seeded, as it's the completed graph of galaxy.stars that it's figuring out the underlying components of.  Did I misidentify the best place for that to happen?

The code in question you're suggesting:
`    def generate_routes(self):
...

        self.generate_base_routes()

        self.logger.info('calculating routes...')
        for star in self.galaxy.stars:`

As I understand that code, `generate_base_routes()` seeds the underlying graph with the edges needed to determine components.  Maybe put `calculate_components()` immediately _after_ `generate_base_routes()`?

As for putting it where I did, I was of the understanding that applying such a filter inside a list comprehension was significantly _faster_ than running through 10+ component checks to each candidate route.

Reply:
For a point of timing reference, I set a full-up Charted Space run going last night under plain Python 3.9 with these changes.
The brute-force-over-underpants doomed-route filtering I'm doing took 304 seconds to run through 5,562,742 routes - 54.7 microseconds per route.

Overall star graph generation took 1 h 47 m 47 s, or 6467 s, as below:
2021-12-12 05:30:08,356 - INFO - generating jumps...
2021-12-12 07:17:55,864 - INFO - base routes: 405456  -  ranges: 5562742
2021-12-12 07:17:55,864 - INFO - calculating routes...
2021-12-12 07:17:56,649 - INFO - Final route count 400700

Doomed-route filtering:
2021-12-12 07:21:13,267 - INFO - filtering routes over 44 components...
2021-12-12 07:26:16,066 - INFO - Filtered route count 5486059

Said run has been running since approx 0528 local, is now 1545 local and 85% of routing gen has just completed

Reply:
That full-space run blew up at 1733 local, during subsector map generation

Reply:
I run the big runs with `--no-subsector-maps` because that will generate a lot of small graphic images. It also breaks in weird places.

Reply:
Had me worried there for a bit.  Any further comment on `calculate_components()` ?

Reply:
Do those most recent changes do what you want?